### ISSUE TITLE
Severe performance regression of the parallel optimizer

### ISSUE THREAD

**Issue by sjrd** [scala-js/scala-js#2864]
Fast-optimizing the testSuite in the 0.6.x branch takes forever; I don't even wait until the end. Reverting back before #2814 fixes the issue.

Our CI jobs seem to take much longer than they used to, as well, so I don't think it is linked to my machine.

At any time, if I ask for a thread dump, all optimizing threads are snapshots with a stack trace similar to the following:

<details>

<summary>Full stack trace</summary>

```
"ForkJoinPool-1-worker-17" #100 daemon prio=5 os_prio=0 tid=0x00007f622400f000 nid=0x2c89 runnable [0x00007f622befb000]
   java.lang.Thread.State: RUNNABLE
	at scala.collection.generic.Growable$$anonfun$$plus$plus$eq$1.apply(Growable.scala:48)
	at scala.collection.generic.Growable$$anonfun$$plus$plus$eq$1.apply(Growable.scala:48)
	at scala.collection.Iterator$class.foreach(Iterator.scala:727)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
	at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at scala.collection.generic.Growable$class.$plus$plus$eq(Growable.scala:48)
	at scala.collection.immutable.VectorBuilder.$plus$plus$eq(Vector.scala:716)
	at scala.collection.parallel.immutable.LazyParVectorCombiner$$anonfun$result$1.apply(ParVector.scala:127)
	at scala.collection.parallel.immutable.LazyParVectorCombiner$$anonfun$result$1.apply(ParVector.scala:126)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
	at scala.collection.parallel.immutable.LazyParVectorCombiner.result(ParVector.scala:126)
	at scala.collection.parallel.immutable.LazyParVectorCombiner.result(ParVector.scala:105)
	at scala.collection.parallel.Combiner$class.resultWithTaskSupport(Combiner.scala:89)
	at scala.collection.parallel.immutable.LazyParVectorCombiner.resultWithTaskSupport(ParVector.scala:105)
	at scala.collection.parallel.ParSeqLike$$anonfun$patch$3.apply(ParSeqLike.scala:218)
	at scala.collection.parallel.ParSeqLike$$anonfun$patch$3.apply(ParSeqLike.scala:218)
	at scala.collection.parallel.ParIterableLike$$anon$12$$anon$4.map(ParIterableLike.scala:306)
	at scala.collection.parallel.ParIterableLike$ResultMapping.leaf(ParIterableLike.scala:962)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply$mcV$sp(Tasks.scala:54)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
	at scala.collection.parallel.Task$class.tryLeaf(Tasks.scala:56)
	at scala.collection.parallel.ParIterableLike$ResultMapping.tryLeaf(ParIterableLike.scala:956)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.compute(Tasks.scala:165)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:514)
	at scala.concurrent.forkjoin.RecursiveAction.exec(RecursiveAction.java:160)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinTask.doJoin(ForkJoinTask.java:341)
	at scala.concurrent.forkjoin.ForkJoinTask.join(ForkJoinTask.java:673)
	at scala.collection.parallel.ForkJoinTasks$WrappedTask$class.sync(Tasks.scala:444)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.sync(Tasks.scala:514)
	at scala.collection.parallel.ForkJoinTasks$class.executeAndWaitResult(Tasks.scala:492)
	at scala.collection.parallel.ForkJoinTaskSupport.executeAndWaitResult(TaskSupport.scala:64)
	at scala.collection.parallel.ParSeqLike$class.patch(ParSeqLike.scala:217)
	at scala.collection.parallel.immutable.ParVector.patch(ParVector.scala:43)
	at scala.collection.parallel.ParSeqLike$class.$colon$plus(ParSeqLike.scala:250)
	at scala.collection.parallel.immutable.ParVector.$colon$plus(ParVector.scala:43)
	at org.scalajs.core.tools.linker.frontend.optimizer.ConcurrencyUtils$AtomicAccOps$.org$scalajs$core$tools$linker$frontend$optimizer$ConcurrencyUtils$AtomicAccOps$$append(ConcurrencyUtils.scala:47)
	at org.scalajs.core.tools.linker.frontend.optimizer.ConcurrencyUtils$AtomicAccOps$.$plus$eq$extension(ConcurrencyUtils.scala:36)
	at org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer$ParMethodImpl.registeredTo(ParIncOptimizer.scala:191)
	at org.scalajs.core.tools.linker.frontend.optimizer.GenIncOptimizer$MethodImpl.registerAskBody(GenIncOptimizer.scala:852)
	at org.scalajs.core.tools.linker.frontend.optimizer.GenIncOptimizer$MethodImpl$Optimizer.getMethodBody(GenIncOptimizer.scala:949)
	at org.scalajs.core.tools.linker.frontend.optimizer.GenIncOptimizer$MethodImpl$Optimizer.getMethodBody(GenIncOptimizer.scala:943)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$inline(OptimizerCore.scala:1844)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformApplyStatic$1.apply(OptimizerCore.scala:1611)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformApplyStatic$1.apply(OptimizerCore.scala:1597)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExprs$1$$anonfun$apply$11.apply(OptimizerCore.scala:748)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExprs$1$$anonfun$apply$11.apply(OptimizerCore.scala:747)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExprs(OptimizerCore.scala:753)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExprs$1.apply(OptimizerCore.scala:747)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExprs$1.apply(OptimizerCore.scala:746)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExpr$1.apply(OptimizerCore.scala:806)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExpr$1.apply(OptimizerCore.scala:787)
	at scala.util.control.TailCalls$TailRec.loop$1(TailCalls.scala:37)
	at scala.util.control.TailCalls$TailRec.result(TailCalls.scala:40)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$trampoline(OptimizerCore.scala:4149)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transform(OptimizerCore.scala:509)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.transformStat(OptimizerCore.scala:276)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$inlineClassConstructorBodyList(OptimizerCore.scala:2308)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$inlineClassConstructorBodyList$2$$anonfun$apply$45.apply(OptimizerCore.scala:2294)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$inlineClassConstructorBodyList$2$$anonfun$apply$45.apply(OptimizerCore.scala:2286)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$inlineClassConstructorBodyList(OptimizerCore.scala:2327)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$inlineClassConstructorBody$1$$anonfun$apply$43.apply(OptimizerCore.scala:2217)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$inlineClassConstructorBody$1$$anonfun$apply$43.apply(OptimizerCore.scala:2209)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$withBindings$1.apply(OptimizerCore.scala:3939)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$withBindings$1.apply(OptimizerCore.scala:3933)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$withNewLocalDefs(OptimizerCore.scala:3966)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$withBindings(OptimizerCore.scala:3940)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$inlineClassConstructorBody$1.apply(OptimizerCore.scala:2218)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$inlineClassConstructorBody$1.apply(OptimizerCore.scala:2189)
	at scala.util.control.TailCalls$TailRec.loop$1(TailCalls.scala:37)
	at scala.util.control.TailCalls$TailRec.result(TailCalls.scala:40)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$trampoline(OptimizerCore.scala:4149)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transform(OptimizerCore.scala:567)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformExpr(OptimizerCore.scala:288)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$11.apply(OptimizerCore.scala:352)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$11.apply(OptimizerCore.scala:317)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformSelectCommon(OptimizerCore.scala:1117)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformSelectCommon$1.apply(OptimizerCore.scala:1050)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformSelectCommon$1.apply(OptimizerCore.scala:1048)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExpr$1.apply(OptimizerCore.scala:806)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExpr$1.apply(OptimizerCore.scala:787)
	at scala.util.control.TailCalls$TailRec.loop$1(TailCalls.scala:37)
	at scala.util.control.TailCalls$TailRec.result(TailCalls.scala:40)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$trampoline(OptimizerCore.scala:4149)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transform(OptimizerCore.scala:355)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.transformStat(OptimizerCore.scala:276)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformList$1(OptimizerCore.scala:728)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.transformBlock(OptimizerCore.scala:735)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transform(OptimizerCore.scala:308)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transform(OptimizerCore.scala:402)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.transformStat(OptimizerCore.scala:276)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformList$1(OptimizerCore.scala:728)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.transformBlock(OptimizerCore.scala:735)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transform(OptimizerCore.scala:308)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.transformIsolatedBody(OptimizerCore.scala:3783)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.liftedTree1$1(OptimizerCore.scala:135)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.optimize(OptimizerCore.scala:134)
	at org.scalajs.core.tools.linker.frontend.optimizer.GenIncOptimizer$MethodImpl.process(GenIncOptimizer.scala:935)
	at org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer$$anonfun$processAllTaggedMethods$2.apply(ParIncOptimizer.scala:88)
	at org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer$$anonfun$processAllTaggedMethods$2.apply(ParIncOptimizer.scala:87)
	at scala.collection.Iterator$class.foreach(Iterator.scala:727)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
	at scala.collection.parallel.ParIterableLike$Foreach.leaf(ParIterableLike.scala:975)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply$mcV$sp(Tasks.scala:54)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
	at scala.collection.parallel.Task$class.tryLeaf(Tasks.scala:56)
	at scala.collection.parallel.ParIterableLike$Foreach.tryLeaf(ParIterableLike.scala:972)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.internal(Tasks.scala:183)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.internal(Tasks.scala:514)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.compute(Tasks.scala:162)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:514)
	at scala.concurrent.forkjoin.RecursiveAction.exec(RecursiveAction.java:160)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

</details>

Note in particular this portion:
```
	at scala.collection.parallel.immutable.ParVector.$colon$plus(ParVector.scala:43)
	at org.scalajs.core.tools.linker.frontend.optimizer.ConcurrencyUtils$AtomicAccOps$.org$scalajs$core$tools$linker$frontend$optimizer$ConcurrencyUtils$AtomicAccOps$$append(ConcurrencyUtils.scala:47)
	at org.scalajs.core.tools.linker.frontend.optimizer.ConcurrencyUtils$AtomicAccOps$.$plus$eq$extension(ConcurrencyUtils.scala:36)
	at org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer$ParMethodImpl.registeredTo(ParIncOptimizer.scala:191)
	at org.scalajs.core.tools.linker.frontend.optimizer.GenIncOptimizer$MethodImpl.registerAskBody(GenIncOptimizer.scala:852)
```
It seems that `ParVector.:+` is an absolute performance killer, that should be avoided at all cost.

**Comment by sjrd** [comment]
@gzm0 Can I ask you to look into this?

**Comment by gzm0** [comment]
Yeah, will do :)

**Comment by sjrd** [comment]
Fixed in #2865.
